programs/conditional_vault/src/instructions/redeem_tokens.rs

```diff
@@ -0,0 +1,127 @@
+use super::*;
+
+impl<'info, 'c: 'info> InteractWithVault<'info> {
+    pub fn validate_redeem_tokens(&self) -> Result<()> {
+        require!(
+            self.question.is_resolved(),
+            VaultError::CantRedeemConditionalTokens
+        );
+
+        Ok(())
+    }
+
+    pub fn handle_redeem_tokens(ctx: Context<'_, '_, 'c, 'info, Self>) -> Result<()> {
+        let accs = &ctx.accounts;
+
+        let (mut conditional_token_mints, mut user_conditional_token_accounts) =
+            Self::get_mints_and_user_token_accounts(&ctx)?;
+
+        // calculate the expected future supplies of the conditional token mints
+        // as current supply - user balance
+        let expected_future_supplies: Vec<u64> = conditional_token_mints
+            .iter()
+            .zip(user_conditional_token_accounts.iter())
+            .map(|(mint, account)| mint.supply - account.amount)
```

Comment:
does this need to be checked, i assume if someone is trying to withdraw `account.amount` > `mint.supply` it will error somewhere along the line

Reply:
I was gonna type "it doesn't hurt to check," but it does because you're adding complexity

I agree this should never be triggered, it's more of a sanity check on myself. my bayesian network tells me it's worth keeping in there :shrug:

Reply:
For sure, just checking in case there was some other reason for it perhaps. Sounds good